### PR TITLE
Add mutator to remove shared cases

### LINKED ISSUE #1290: Add Mutator that removes cases from switch-Statements
I would like to suggest a new mutator, that removes `case`s from `switch`-Statements. I'm not 100% sure, but this is probably only needed, if the `case` has an empty body or if there are multiple `case`s with the same body. In the first case, there is no code that can be otherwise mutated, in the second case it is possible that not all `case`s are actually covered by tests.

Example:
Assume code like this:
```
switch ($value) {
  case 'a':
    doSomething();
    break;
  case 'b':
  case 'c':
  default:
    doSomethingElse();
    break;
}
```
If all tests only execute this code with `$value = 'a'` or `$value = 'b'` infection should report an escaped mutant if `case 'c':` or `default:` is removed.

### PR DESCRIPTION
**Pull request by Khartir**
This PR:
- [x] Adds a new mutator to remove `case`s from `switch`-statments that share a body
- [x] Covered by tests
- [x] Doc PR: https://github.com/infection/site/pull/173

Closes #1290

Code like this:
```
switch ($value) {
  case 'a':
    doSomething();
    break;
  case 'b':
  case 'c':
  default:
    doSomethingElse();
    break;
}
```
Creates the following mutants:
```diff
switch ($value) {
  case 'a':
    doSomething();
    break;
-  case 'b':
  case 'c':
  default:
    doSomethingElse();
    break;
}
```
```diff
switch ($value) {
  case 'a':
    doSomething();
    break;
  case 'b':
-  case 'c':
  default:
    doSomethingElse();
    break;
}
```
```diff
switch ($value) {
  case 'a':
    doSomething();
    break;
  case 'b':
  case 'c':
-  default:
    doSomethingElse();
    break;
}
```

### CODE REVIEW

File: src/Mutator/Removal/SharedCaseRemoval.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+     *
+     * @return iterable<Node\Stmt\Switch_>
+     */
+    public function mutate(Node $node): iterable
+    {
+        $lastWasEmpty = false;
+
+        foreach ($node->cases as $i => $case) {
+            if ($case->stmts === []) {
+                $lastWasEmpty = true;
```
Comment (maks-rafalko) [inline src/Mutator/Removal/SharedCaseRemoval.php:86]:
```suggestion
                $previousCaseWasEmpty = true;
```

what do you think? As for me, last is the "the last element in the array". Up to you :)

**Review by maks-rafalko — COMMENTED** [review]
Thanks for working on this new Mutator! Looks good to me except minor comments above.

Question: why do we remove only *shared* cases? Why not to remove case by case?

**Comment by Khartir** [comment]
@maks-rafalko I only removed shared cases because there is no way to see they are missing tests by looking at the code-coverage. 
Removing a "normal" case would be like removing an entire `if`. I assumed that this would be better detected by the existing mutators mutating the code inside the `case`.
Which brings up the edge case of
```php
case 'a':
  break;
```
I'm unsure about this. If you want this to remove all `case`s (including their body) I can do so. There should then probably also be a mutator to remove `if` and `else`.

What do you think?

Regarding the suggested changes: I'll probably look into it over the weekend.

**Comment by maks-rafalko** [comment]
I'm ok with removing only shared cases :) Thanks, looking forward to getting it merged!

File: src/Mutator/Removal/SharedCaseRemoval.php (hunk 1 above)
Comment (Khartir) [reply src/Mutator/Removal/SharedCaseRemoval.php]: Sounds good, fixed.

File: src/Mutator/Removal/SharedCaseRemoval.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                $cases = $node->cases;
+                unset($cases[$i]);
+
+                yield new Node\Stmt\Switch_(
+                    $node->cond,
+                    $cases,
+                    $node->getAttributes()
+                );
+            } elseif ($lastWasEmpty) {
```
[the remark this answers is not in the document]
Comment (Khartir) [reply src/Mutator/Removal/SharedCaseRemoval.php]: Fixed.

File: tests/phpunit/Mutator/Removal/SharedCaseRemovalTest.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    public function mutationsProvider(): iterable
+    {
+        yield 'It does not mutate single cases with a body' => [
+            '<?php switch(true) {case true: $a = [];break; case false: $a = [];break;};',
+        ];
+
+        yield 'It removes cases that share a body' => [
+            '<?php switch(true) {case true: case false: $a = [];break; case null: $a = []; break;};',
```
[the remark this answers is not in the document]
Comment (Khartir) [reply tests/phpunit/Mutator/Removal/SharedCaseRemovalTest.php]: Fixed.

**Review by sanmai — APPROVED** [review]

**Review by maks-rafalko — APPROVED** [review]
Thank you @Khartir!

**Pull request merged**